### ISSUE TITLE
Epic: Cryptography tasks prior to mainnet

### ISSUE THREAD

**Issue by redshiftzero** [penumbra-zone/penumbra#1825]
This is a tracking ticket for the core protocol cryptography tasks that need to be completed prior to mainnet. 

# Finalize core cryptography

The current status as of filing this issue is:
* We have finalized the crypto primitives (decaf377, poseidon377)
* We have finalized the key/address derivation
* We have integrated the pieces we need to support FMD in the future (FMD clues are in transactions, etc.)
* We lifted the memo up to the tx level and added another layer of key wrapping to enable each output in the tx to decrypt the memo

Remaining tasks prior to mainnet:
- [x] https://github.com/penumbra-zone/penumbra/issues/1688
- [x] https://github.com/penumbra-zone/penumbra/issues/1716
- [x] https://github.com/penumbra-zone/penumbra/issues/1747
- [x] https://github.com/penumbra-zone/penumbra/issues/1746
- [x] https://github.com/penumbra-zone/penumbra/issues/1962

# ZK Proofs

The current status of this work is that we've implemented:
* low level r1cs gadgets for decaf377 and poseidon377
* high level r1cs gadgets for all the functionality we need for the spend and output proofs: value commitment integrity, merkle path integrity, nullifier integrity, ephemeral public key integrity, checking generator points are not identity, randomized verification key integrity.
* ZK spend and output proofs are implemented (but not in use in testnets yet)

Tasks prior to mainnet that can be done immediately:
- [x] https://github.com/penumbra-zone/penumbra/issues/711
- [x] https://github.com/penumbra-zone/penumbra/issues/717 (will be closed by https://github.com/penumbra-zone/penumbra/pull/1970)
- [x] Replace transparent spend and output proofs with their ZK versions (M) (will be closed by #1923)
- [x] Benchmarking: How many constraints per high-level gadget? (results [here](https://github.com/penumbra-zone/penumbra/issues/1825#issuecomment-1377477140))
- [x] [Fix intermittent verification issues uncovered during #1923](https://github.com/penumbra-zone/penumbra/issues/2041)
- [x] https://github.com/penumbra-zone/penumbra/issues/2040
- [x] https://github.com/penumbra-zone/penumbra/issues/2056
- [x] https://github.com/penumbra-zone/penumbra/issues/1824
- [x] #2152 (closed by #2272)
- [x] Fixed point arithmetic: will be closed by #2389 
- [x] https://github.com/penumbra-zone/penumbra/issues/1093 (closed by #2618)
- [x] https://github.com/penumbra-zone/penumbra/issues/1092 (closed by #2083)
- [x] https://github.com/penumbra-zone/penumbra/issues/1095
- [x] https://github.com/penumbra-zone/penumbra/issues/2004

## Documentation

  - [x] Output: closed by https://github.com/penumbra-zone/penumbra/pull/2112
  - [x] Spend: closed by https://github.com/penumbra-zone/penumbra/pull/2347
  - [x] Swap: closed by https://github.com/penumbra-zone/penumbra/pull/2445
  - [x] SwapClaim: closed by #2650
  - [x] Undelegate claim: closed by https://github.com/penumbra-zone/penumbra/pull/2447
  - [x] Delegator Vote: closed by #2355

## Benchmarking / Optimization / Assurance

- [x] https://github.com/penumbra-zone/penumbra/issues/2350
- [x] Add Rust benchmarks to repo so it's easier to generate data like the one-off benchmarking we did in the comment below - https://github.com/penumbra-zone/penumbra/issues/2039 (important as we start doing optimization)
- [x] Optimization: https://github.com/penumbra-zone/poseidon377/issues/30
- [x] Further optimization?

## Proof parameters setup

- [x] https://github.com/penumbra-zone/penumbra/issues/712 - needs fleshing out

**Comment by redshiftzero** [comment]
Here are the results of benchmarking circuit costs of our gadgets as of commit https://github.com/penumbra-zone/penumbra/commit/0db4a2be63ff444aa5e481cf82e6eb22e6653500:

Gadgets:
- Witnessing a single decaf377 element: 1392

The numbers below include the constraint cost of witnessing the required instance variables: 
- `diversified_basepoint_not_identity` (requires 1 `Element` instance variable): 1396
- `ephemeral_public_key_integrity` (requires 2 `Element` instance variable, plus 1 scalar): 4982
- `value_commitment_integrity` (requires 1 `Element` instance variable, 1 Fq, plus 2 scalar): 3376
- `note_commitment_integrity` (requires 6 Fq, 1 `Element`): 3201
- Verifying merkle path (requires 2 Fq, 1 merkle path variable (72 Fqs)): 8631
- `rk_integrity` (1 `Element`, 1 scalar, 1 Fq): 5689
- `nullifier_integrity` (4 Fq): 311
- `ak_not_identity` (1 `Element`): 1396

**Comment by redshiftzero** [comment]
Closing this as the cryptography tasks described here are completed, there is some orchestration work for the ceremony (cryptography parts were tracked in #712) remaining that will be tracked elsewhere. Custody work will also be tracked in a separate ticket.
